### ISSUE TITLE
Tup no longer works with extra outputs in macros when used in foreach

### ISSUE THREAD

**Issue by andrewwjensen** [gittup/tup#246]
I have a tup macro with extra inputs. This functionality is broken in recent tup versions if that macro is used with a `foreach`. Here is a simple example:

main.cc:
```c++
    int main(){return 0;}
```
foo.cc:
```c++
    int foo() {return 10;}
```
bar.cc:
```c++
    int bar() {return 20;}
```
Tupfile:
```
    !compile = |> g++ -c %f --coverage -o %o |> %B.o | %B.gcno
    : foreach *.cc |> !compile |>
```

```
bash$ tup --version
tup v0.7.3-35-g0ed01a5
bash$ tup init
.tup repository initialized.
bash$ tup
[ tup ] [0.000s] Scanning filesystem...
[ tup ] [0.001s] Reading in new environment variables...
[ tup ] [0.001s] Parsing Tupfiles...
 1) [0.002s] .
 [ ] 100%
[ tup ] [0.015s] No files to delete.                                                                        
[ tup ] [0.015s] Generating .gitignore files...
[ tup ] [0.016s] Executing Commands...
* 1) g++ -c foo.cc --coverage -o foo.o                                                                      
 *** tup errors ***
tup error: File '/home/ajensen/tupdbg/foo.gcno' was written to, but is not in .tup/db. You probably should specify it as an output
 *** Command ID=14 ran successfully, but tup failed to save the dependencies.
* 2) g++ -c main.cc --coverage -o main.o                                                                    
 *** tup errors ***
tup error: File '/home/ajensen/tupdbg/main.gcno' was written to, but is not in .tup/db. You probably should specify it as an output
 *** Command ID=16 ran successfully, but tup failed to save the dependencies.
 3) [0.025s] g++ -c bar.cc --coverage -o bar.o                                                              
 [   ] 100%
 *** tup: 2 jobs failed.
```

Note that it compiles the alphabetically first file (`bar.cc`) of the `foreach` fine, but all subsequent files have errors. It's like the `%B.gcno` extra output is only getting applied to the first file matched by the `foreach`.

If I change the Tupfile to have the `%B.o | %B.gcno` on the `foreach` line instead of the `!compile` macro, it works. Even if I just move the extra output portion (`| %B.gcno`) to the `foreach` line, leaving the `%B.o` in the macro, it works.

Note also that if I `tup init` with tup built from commit 3def53adae15f54083b1c99b17d82feabbef4501 or earlier, then I can use any version of tup (including the one that fails above) to build and it works. However, if I delete the .tup directory and `tup init` with the later "bad" version, it fails. It will even still fail to build if I then try a "good" tup version until I `tup init` with the good version.

The first version that fails to "init" properly is: b80e17f29e6ae81041c240c2d1ec320a1ceddd5e

The latest good version is: 3def53adae15f54083b1c99b17d82feabbef4501

**Comment by xguerin** [comment]
I have a similar issue, due to the same commit b80e17f29e6ae81041c240c2d1ec320a1ceddd5e, involving macros with dependency groups and variants. I uploaded a regression test here: https://github.com/xguerin/tup-macro-bug.

**Comment by gittup** [comment]
Thanks for the test case! The groups issue should be fixed by 95f32fbee16332ad7b921d6ca50883e52fd6fb05. The original issue with extra outputs was fixed by 28beb20626ba5459df9a5e5c25ece6396b4214a3 I believe.

**Comment by xguerin** [comment]
Awesome, thanks !!
